scipy/spatial/distance.py

```diff
@@ -802,12 +805,21 @@ def hamming(u, v, w=None):
     0.33333333333333331
     >>> distance.hamming([1, 0, 0], [3, 0, 0])
     0.33333333333333331
+    >>> distance.hamming("hello", "world")
+    0.8
+    >>> distance.hamming("same", "same")
+    0.0
 
     """
+    if isinstance(u, str) and isinstance(v, str):
+        u = list(u)
+        v = list(v)
```

Comment:
I see this PR has been sitting for a while. There was some contention in the original issue proposing this change because NumPy by default treats the strings as scalars.

While I agree that the the default NumPy behavior is not "in the spirit" of what the hamming distance is typically used for (how many elementwise differences are there between two sequences of string or other types and *not* the differences between the composite scalars), we may need to exercise some caution if we want to have a change in behavior.

@WarrenWeckesser what about adding a new argument? I guess the question here is whether the burden on users is really so high that we can't for example just improve the docstring on this matter (i.e., suggest list conversion for elementwise string comparisons) vs. expanding the code/available options?

Reply:
I absolutely agree with the caution. A deprecation warning to alert the user to the change would definitely be in order. Something like `"The hamming distance will compute the character wise comparison of strings in future versions of scipy. To retain the current behavior of the string comparison the strings need to be array elements. for further information about the change see [link to this PR or some or the docs]."`

Just because numpy treats strings as scalars, does not mean that putting the string into an array is the right thing to do, because the string argument violates the idea of the hamming distance computation by not providing an array in the first place. So the question is not how numpy treats strings, but how the hamming distance should be recovering from the contract violation. My argument is simply that the current handling is not intuitive and thus a suboptimal API.

Maybe it would be even better to just throw a ValueError if the argument are strings to avoid making assumptions altogether?

Reply:
BTW: the jaccard distance already has a FutureWarning in the spirit of this PR:

```python
>>> import scipy
>>> scipy.__version__
'1.6.2'
>>> from scipy.spatial.distance import jaccard
>>> jaccard("aa", "ab")
FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
1.0
```

In the spirit of a uniform user experience within scipy, there probably should be a consistent handling of the case where strings are supplied to a distance metric. In this case this would mean to add a FutureWarning for a while and then change the behavior as proposed here.